master/internal/api_project.go

```diff
@@ -197,71 +197,65 @@ func (a *apiServer) getProjectColumnsByID(
 		}
 	}
 	summaryMetrics := []struct {
-		ID             int
-		SummaryMetrics model.JSONObj
+		MetricName string
+		JSONPath   string
+		MetricType string
+		Count      int32
 	}{}
 
 	if len(trialIDs) > 0 {
-		trialsQuery := db.Bun().NewSelect().
-			Column("id", "summary_metrics").
-			Table("trials").
-			Where("id IN (?)", bun.In(trialIDs)).
-			Order("id")
+		subQuery := db.BunSelectMetricGroupNames().ColumnExpr(
+			`summary_metrics->jsonb_object_keys(summary_metrics)->
+		jsonb_object_keys(summary_metrics->jsonb_object_keys(summary_metrics))->>'type'
+		AS metric_type`).
+			Where("id IN (?)", bun.In(trialIDs)).Distinct()
+		trialsQuery := db.Bun().NewSelect().TableExpr("(?) AS stats", subQuery).
+			ColumnExpr("*").ColumnExpr(
+			"ROW_NUMBER() OVER(PARTITION BY json_path, metric_name order by metric_type) AS count").
+			Order("json_path").Order("metric_name")
 		err = trialsQuery.Scan(ctx, &summaryMetrics)
 		if err != nil {
 			return nil, err
 		}
 	}
 
-	summaryMetricsSet := make(map[string]struct{})
-	for _, trial := range summaryMetrics {
-		if rawMetrics, ok := trial.SummaryMetrics["avg_metrics"]; ok {
-			// iterate in order
-			metrics := rawMetrics.(map[string]interface{})
-			metricsKeys := make([]string, 0, len(metrics))
-			for metricKey := range metrics {
-				metricsKeys = append(metricsKeys, metricKey)
+	for _, stats := range summaryMetrics {
+		// If there are multiple metrics with the same group and name, report the type as unspecified.
+		if stats.Count > 1 {
+			columns[len(columns)-1].Type = projectv1.ColumnType_COLUMN_TYPE_UNSPECIFIED
+		}
+		switch stats.JSONPath {
+		case metricGroupValidation:
+			columnType := parseMetricsType(stats.MetricType)
+			columns = append(columns, &projectv1.ProjectColumn{
+				Column:   fmt.Sprintf("validation.%s", stats.MetricName),
+				Location: projectv1.LocationType_LOCATION_TYPE_VALIDATIONS,
+				Type:     columnType,
+			})
+		default:
+			columnType := parseMetricsType(stats.MetricType)
+			columnPrefix := stats.JSONPath
```

Comment:
i'm new to this bit of code; why are training and validation metrics handled so differently?